### PR TITLE
Reader Onboarding: Update the subscriptions list behind the modal

### PR DESCRIPTION
**Pull request by allilevine**
Fixes https://github.com/Automattic/loop/issues/204

## Proposed Changes

* Pass a `handleFollowToggle` function from the subscribe modal, that re-requests follows to update the "Subscriptions" list.

https://github.com/user-attachments/assets/1d9b69db-75e0-4408-b21a-f1e42485ee3e

## Why are these changes being made?

* To improve the Reader Onboarding user experience.

## Testing Instructions

* As a new user with a verified email, go to /read?flags=reader/onboarding
* Click "Select some of your interests" to begin onboarding, choose interests, and click Continue.
* On the next step, Subscribe to a site. Hit cancel or continue.
* Check that the "Subscriptions" list on the right includes the site.
* Clear your onboarding completed preference and go back to the "Discover and subscribe to sites you'll love" step.
* Unsubscribe from the site. Hit cancel or continue.
* Check that the site is no longer listed in Subscriptions.

Known existing issue: the Subscribe/Subscribed button is slightly glitchy.

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

File: client/blocks/reader-subscription-list-item/index.jsx [hunk 1]
```diff
@@ -220,7 +221,7 @@ function ReaderSubscriptionListItem( {
 					feedId={ feedId }
 					siteId={ siteId }
 					railcar={ railcar }
-					onFollowToggle={ openSuggestedFollowsModal }
+					onFollowToggle={ onFollowToggle || openSuggestedFollowsModal }
```
Comment (allilevine) [inline client/blocks/reader-subscription-list-item/index.jsx]: I haven't dug into if this causes issues in other components yet.

File: client/reader/onboarding/subscribe-modal/index.tsx [hunk 2]
```diff
@@ -208,6 +209,38 @@ const SubscribeModal: React.FC< SubscribeModalProps > = ( { isOpen, onClose } )
...
+
+			// Exit early if the follow state already matches what we want.
+			if ( following === isFollowingSite( site ) ) {
+				return;
+			}
+
+			// Maximum number of retries
+			const MAX_RETRIES = 3;
+
+			for ( let attempt = 0; attempt < MAX_RETRIES; attempt++ ) {
```
Comment (allilevine) [inline client/reader/onboarding/subscribe-modal/index.tsx:227]: The follow or unfollow is happening async so I'm checking if it's completed before re-fetching. This is where the follow happens:

https://github.com/Automattic/wp-calypso/blob/update/reader-onboarding-subscriptions/client/blocks/follow-button/index.jsx#L52

File: client/blocks/reader-subscription-list-item/index.jsx (hunk 1 above)
Comment (allilevine) [reply client/blocks/reader-subscription-list-item/index.jsx]: I added a check for `disableSuggestedFollows`.

**Review by DustyReagan — APPROVED** [review]
This is working nicely for me!

**Pull request merged**